src/adaptive_threshold.jl

```diff
@@ -1,97 +1,137 @@
+
 # Options
-Various options for the parameters of this function are described in more detail
-below.
+
+Various options for the parameters of `AdaptiveThreshold`, `binarize` and
+`binarize!` are described in more detail below.
 
 ## Choices for `percentage`
-You can specify an integer for the `percentage` (denoted by ``t`` in the
-publication) which must be between 0 and 100. If left unspecified a default
-value of 15 is utilised.
+
+You can specify an integer for the `percentage` (denoted by `t` in [1])
+which must be between 0 and 100. Default: 15
 
 ## Choices for `window_size`
-The argument `window_size` (denoted by ``s`` in the publication)  specifies the
-size of pixel's square neighbourhood which must be greater than zero. A
-recommended size is the integer value which is closest to 1/8 of the average of
-the width and height. You can use the convenience function
-`recommend_size(::AbstractArray{T,2})` to obtain this suggested value.
-If left unspecified, a default value of 32 is utilised.
 
-#Example
+The argument `window_size` (denoted by `s` in [1]) specifies the size of
+pixel's square neighbourhood which must be greater than zero.
+
+The default value is the integer which is closest to 1/8 of the average of
+the width and height of `img`.
+
+!!! info
+
+    `window_size` is a keyword argument in [`binarize`](@ref) and [`binarize!`](@ref)
+
+# Examples
 
 ```julia
 using TestImages
 
 img = testimage("cameraman")
-s = recommend_size(img)
-binarize(AdaptiveThreshold(percentage = 15, window_size = s), img)
+f = AdaptiveThreshold()
+
+img₀₁_1 = binarize(img, f)
+img₀₁_2 = binarize(img, f, window_size=16)
 ```
 
+See also [`binarize!`](@ref) for in-place operation.
+
 # References
-1. Bradley, D. (2007). Adaptive Thresholding using Integral Image. *Journal of Graphic Tools*, 12(2), pp.13-21. [doi:10.1080/2151237x.2007.10129236](https://doi.org/10.1080/2151237x.2007.10129236)
+
+[1] Bradley, D. (2007). Adaptive Thresholding using Integral Image. *Journal of Graphic Tools*, 12(2), pp.13-21. [doi:10.1080/2151237x.2007.10129236](https://doi.org/10.1080/2151237x.2007.10129236)
+
 """
-function binarize(algorithm::AdaptiveThreshold, img::AbstractArray{T,2}) where T <: Colorant
-    binarize(algorithm, Gray.(img))
+struct AdaptiveThreshold <: AbstractImageBinarizationAlgorithm
+    percentage::Float64
+
+    function AdaptiveThreshold(percentage)
+        (percentage < 0 || percentage > 100) && throw(ArgumentError("percentage should be ∈ [0, 100]."))
+        new(percentage)
+    end
+end
+
+function AdaptiveThreshold(; percentage::Real = 15, window_size=nothing)
+    if window_size === nothing
+        return AdaptiveThreshold(percentage)
+    else
+        # deprecate window_size
+        return AdaptiveThreshold(window_size, percentage)
+    end
 end
 
-function binarize(algorithm::AdaptiveThreshold, img::AbstractArray{T,2}) where T <: Gray
-    s = algorithm.window_size
-    t = algorithm.percentage
-    if s < 0 || t < 0 || t > 100
```

Comment:
As a reminder to myself:

we need to check whether the given `window_size` is valid, i.e., `>=0` for `AdaptiveThreshold`, `Niblack`, `Sauvola`.

There was one for `AdaptiveThreshold` but I accidentally removed them while moving `window_size` out of the constructor.